### PR TITLE
libobs: Add HLSL annotation parsing

### PR DESCRIPTION
**Pull request by Andersama**
Note: Dependent on #1319
Gives the ability to retrieve param annotations. Blocks wrapped in <>
following a parameter.

For example:
```c
float slider < float max_value = 10.0; float min_value = 0.0; >;
```
These blocks are not for shading purposes but to help describe the shader's gui
as in the example above.

Adds graphics api functions for retrieving annotations and their default values:
```c
gs_eparam_t *gs_param_get_annotation_by_idx(const gs_eparam_t *param,
		size_t annotation);

gs_eparam_t *gs_param_get_annotation_by_name(const gs_eparam_t *param,
		const char *name);

void *gs_effect_get_val(gs_eparam_t *param);

void *gs_effect_get_default_val(gs_eparam_t *param);
```

Example usage:
[shader-filter](https://github.com/andersama/obs-studio/tree/shader-filter-eval)

### CODE REVIEW

File: libobs/graphics/effect-parser.c [hunk 1]
```diff
@@ -356,9 +525,16 @@ static void ep_parse_technique(struct effect_parser *ep)
 
 	if (cf_next_name(&ep->cfp, &ept.name, "name", ";") != PARSE_SUCCESS)
 		goto error;
-	if (cf_next_token_should_be(&ep->cfp, "{", ";", NULL) != PARSE_SUCCESS)
+	if (!cf_next_valid_token(&ep->cfp)) return PARSE_EOF;
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: You've returned a non-void here which is causing a compile failure on macOS.

File: libobs/graphics/effect-parser.c [hunk 2]
```diff
@@ -356,9 +525,16 @@ static void ep_parse_technique(struct effect_parser *ep)
...
+	if (cf_token_is(&ep->cfp, "<") && !ep_parse_technique_annotations(ep, &ept))
 		goto error;
+	if (!cf_token_is(&ep->cfp, "{")) {
+		if (!cf_go_to_token(&ep->cfp, ";", NULL))
+			return PARSE_EOF;
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: You've returned a non-void here which is causing a compile failure on macOS.

File: libobs/graphics/effect-parser.c [hunk 3]
```diff
@@ -262,6 +277,155 @@ static void ep_parse_struct(struct effect_parser *ep)
...
+
+		case PARSE_EOF:
+			ep_param_free(&var);
+			goto error;
+		}
+
+		if (do_break)
+			break;
+
+		//da_push_back(annotations, &var);
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Removed commented line if no longer needed.

File: libobs/graphics/effect-parser.c [hunk 4]
```diff
@@ -1339,6 +1540,97 @@ static void ep_makeshaderstring(struct effect_parser *ep,
...
+			param->type = GS_SHADER_PARAM_TEXTURE;
+	}
+}
+
+static void ep_compile_param_annotations(struct ep_param *ep_param_input,
+	struct gs_effect_param *gs_effect_input, struct effect_parser *ep)
+{
+	ep_compile_annotations(&(ep_param_input->annotations.da),
+		&(gs_effect_input->annotations.da), ep);
+	/*
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Remove commented out code section if no longer needed.

File: libobs/graphics/effect.c [hunk 5]
```diff
@@ -286,6 +286,107 @@ gs_eparam_t *gs_effect_get_param_by_name(const gs_effect_t *effect,
...
 
+gs_eparam_t *gs_param_get_annotation_by_idx(const gs_eparam_t *param,
+	size_t annotation)
+{
+	if (!param) return NULL;
+	struct gs_effect_param *params = param->annotations.array;
+	if (annotation > param->annotations.num)
+		return NULL;
+
+	return params+annotation;
```
Comment (RytoEX) [inline libobs/graphics/effect.c]: Style:  Put a space on each side of a math operator.

File: libobs/graphics/graphics.h [hunk 6]
```diff
@@ -368,13 +369,32 @@ EXPORT bool gs_technique_begin_pass(gs_technique_t *technique, size_t pass);
...
+		const char *name);
+EXPORT gs_eparam_t *gs_technique_get_annotation_by_idx(
+	const gs_technique_t *technique, size_t annotation);
+EXPORT gs_eparam_t *gs_technique_get_annotation_by_name(
+	const gs_technique_t *technique, const char *name);
+EXPORT gs_eparam_t *gs_pass_get_annotation_by_idx(const gs_epass_t *pass,
+		size_t annotation);
+EXPORT gs_eparam_t *gs_pass_get_annotation_by_name(const gs_epass_t *pass,
+		const char *name);
+/*
```
Comment (RytoEX) [inline libobs/graphics/graphics.h]: Remove commented out code section if no longer needed.

**Comment by Andersama** [comment]
Any clues as to the mac build failure?

**Comment by Andersama** [comment]
Jim described what that function was supposed to do, so I should actually probably close that other pr*, I'll be happy to split the pr into two parts.

**Comment by Andersama** [comment]
This pull request can be considered finished, memory leaks among other problems were solved.

File: docs/sphinx/reference-libobs-graphics-effects.rst [hunk 7]
```diff
@@ -154,6 +154,36 @@ HLSL format.
 
 ---------------------
 
+.. function:: size_t gs_param_get_num_annotations(const gs_eparam_t *param)
+
+   Gets the number of annotations associated with the parameter.
+
+   :param param: Param object
```
Comment (RytoEX) [inline docs/sphinx/reference-libobs-graphics-effects.rst]: Style:  Align "Param object" with with the lines below.  See examples throughout file.

File: libobs/graphics/effect-parser.c [hunk 8]
```diff
@@ -21,6 +21,39 @@
 #include "effect-parser.h"
 #include "effect.h"
 
+static inline bool ep_parse_param_assign(struct effect_parser *ep,
+	struct ep_param *param);
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Style:  Tab twice for continued lines.

File: libobs/graphics/effect-parser.c [hunk 9]
```diff
@@ -92,6 +125,18 @@ static inline struct ep_param *ep_getparam(struct effect_parser *ep,
 	return NULL;
 }
 
+static inline struct ep_param *ep_getannotation(struct ep_param *param,
+	const char *name)
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Style:  Tab twice for continued lines.

File: libobs/graphics/effect-parser.c [hunk 10]
```diff
@@ -262,6 +307,144 @@ static void ep_parse_struct(struct effect_parser *ep)
 	ep_struct_free(&eps);
 }
 
+static inline int ep_parse_param_annotation_var(struct effect_parser *ep,
+	struct ep_param *var)
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Style:  Tab twice for continued lines.

File: libobs/graphics/effect-parser.c [hunk 11]
```diff
@@ -262,6 +307,144 @@ static void ep_parse_struct(struct effect_parser *ep)
...
+		cf_adderror_expecting(&ep->cfp, "variable name");
+		return PARSE_UNEXPECTED_CONTINUE;
+	}
+	if (cf_token_is(&ep->cfp, ">")) {
+		cf_adderror_expecting(&ep->cfp, "variable name");
+		return PARSE_UNEXPECTED_BREAK;
+	}
+
+	code = cf_token_is_type(&ep->cfp, CFTOKEN_NAME, "variable name",
+		";");
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Style:  Tab twice for continued lines.

File: libobs/graphics/effect-parser.c [hunk 12]
```diff
@@ -262,6 +307,144 @@ static void ep_parse_struct(struct effect_parser *ep)
...
+	if (code != PARSE_SUCCESS)
+		return code;
+
+	bfree(var->name);
+	cf_copy_token(&ep->cfp, &var->name);
+
+	/* -------------------------------------- */
+	/* variable mapping if any (POSITION, TEXCOORD, etc) */
+
+	if (!cf_next_valid_token(&ep->cfp)) return PARSE_EOF;
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Style:  Avoid combining multiple statements on a single line.

File: libobs/graphics/effect-parser.c [hunk 13]
```diff
@@ -262,6 +307,144 @@ static void ep_parse_struct(struct effect_parser *ep)
...
+			return PARSE_EOF;
+		}
+		return PARSE_CONTINUE;
+	}
+
+	return PARSE_SUCCESS;
+}
+
+static int ep_parse_annotations(struct effect_parser *ep,
+	struct darray *annotations)
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Style:  Tab twice for continued lines.

File: libobs/graphics/effect-parser.h [hunk 14]
```diff
@@ -99,6 +102,9 @@ static inline void ep_param_free(struct ep_param *epp)
 	bfree(epp->name);
 	da_free(epp->default_val);
 	da_free(epp->properties);
+	for (size_t i = 0; i < epp->annotations.num; i++)
+		ep_param_free(epp->annotations.array + i);
+	da_free(epp->annotations);
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.h]: Add a blank line above this one.

File: libobs/graphics/effect-parser.c (hunk 12 above)
Comment (Andersama) [reply libobs/graphics/effect-parser.c]: There were a few of these from the original code, wasn't sure if I should follow along w/ that at the time. I'll go over my functions again.

Comment (RytoEX) [reply libobs/graphics/effect-parser.c]: There are some cases where I might consider it disruptive to stray from existing patterns within a particular file/subproject/submodule (e.g., alignments, indents, naming conventions).  This is not one of them.  I'm not saying you need to change all of the previously existing instances, just don't introduce new ones with that pattern.

File: libobs/graphics/effect-parser.c [hunk 15]
```diff
@@ -262,6 +307,146 @@ static void ep_parse_struct(struct effect_parser *ep)
...
+	/* get annotation variables */
+	while (true) {
+		bool do_break = false;
+		struct ep_param var;
+
+		ep_param_init(&var, bstrdup(""), bstrdup(""), false, false,
+				false);
+
+		switch (ep_parse_param_annotation_var(ep, &var)) {
+
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Remove this blank line.

File: libobs/graphics/effect-parser.c [hunk 16]
```diff
@@ -262,6 +307,146 @@ static void ep_parse_struct(struct effect_parser *ep)
...
+		goto error;
+
+	return true;
+
+error:
+	return false;
+}
+
+static int ep_parse_param_annotations(struct effect_parser *ep,
+	struct ep_param *param)
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Style:  Tab twice.

File: libobs/graphics/effect-parser.c [hunk 17]
```diff
@@ -328,7 +513,8 @@ static int ep_parse_pass(struct effect_parser *ep, struct ep_pass *pass)
 
 	if (!cf_token_is(&ep->cfp, "{")) {
 		pass->name = bstrdup_n(ep->cfp.cur_token->str.array,
-		                        ep->cfp.cur_token->str.len);
+			ep->cfp.cur_token->str.len);
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: This seems like an unrelated style change to an existing line.  While I _would_ eventually like to bring style consistency to the existing codebase, it's not generally hidden within an existing commit for changing functionality.

File: libobs/graphics/effect-parser.c [hunk 18]
```diff
@@ -356,9 +542,18 @@ static void ep_parse_technique(struct effect_parser *ep)
 
 	if (cf_next_name(&ep->cfp, &ept.name, "name", ";") != PARSE_SUCCESS)
 		goto error;
-	if (cf_next_token_should_be(&ep->cfp, "{", ";", NULL) != PARSE_SUCCESS)
-		goto error;
 
+	if (!cf_next_valid_token(&ep->cfp)) return;
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Style:  Don't write multiple statements on one line.

File: libobs/graphics/effect-parser.c [hunk 19]
```diff
@@ -756,6 +951,31 @@ static inline int ep_parse_param_assign_texture(struct effect_parser *ep,
 	return PARSE_SUCCESS;
 }
 
+static inline int ep_parse_param_assign_string(struct effect_parser *ep,
+	struct ep_param *param)
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Style:  Tab twice.

File: libobs/graphics/effect-parser.c [hunk 20]
```diff
@@ -756,6 +951,31 @@ static inline int ep_parse_param_assign_texture(struct effect_parser *ep,
...
+{
+	int code;
+	char *str = NULL;
+
+	if (!cf_next_valid_token(&ep->cfp))
+		return PARSE_EOF;
+
+	code = cf_token_is_type(&ep->cfp, CFTOKEN_STRING,
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Style:  This does not need to be line-broken.

File: libobs/graphics/effect-parser.c [hunk 21]
```diff
@@ -756,6 +951,31 @@ static inline int ep_parse_param_assign_texture(struct effect_parser *ep,
...
+		"string", ";");
+	if (code != PARSE_SUCCESS)
+		return code;
+
+	str = cf_literal_to_str(ep->cfp.cur_token->str.array,
+		ep->cfp.cur_token->str.len);
```
Comment (RytoEX) [inline libobs/graphics/effect-parser.c]: Style:  Tab twice.

File: libobs/graphics/effect-parser.c [hunk 22]
```diff
@@ -1013,7 +1259,6 @@ bool ep_parse(struct effect_parser *ep, gs_effect_t *effect,
 			cf_adderror(&ep->cfp, "Unexpected code segment",
 					LEX_ERROR, NULL, NULL, NULL);
 			cf_pass_pair(&ep->cfp, '{', '}');
-
```
Comment (jp9000) [inline libobs/graphics/effect-parser.c]: Unnecessary change along with the 3 others above this.

File: libobs/graphics/effect-parser.h [hunk 23]
```diff
@@ -126,8 +133,8 @@ static inline void ep_struct_init(struct ep_struct *eps)
 static inline void ep_struct_free(struct ep_struct *eps)
 {
 	size_t i;
-
 	bfree(eps->name);
+
```
Comment (jp9000) [inline libobs/graphics/effect-parser.h]: Unnecessary change here as well.

File: libobs/graphics/effect-parser.c [hunk 24]
```diff
@@ -21,6 +21,39 @@
 #include "effect-parser.h"
 #include "effect.h"
 
+static inline bool ep_parse_param_assign(struct effect_parser *ep,
+		struct ep_param *param);
+
+enum gs_shader_param_type get_effect_param_type(const char *type)
```
[the remark this answers is not in the document]
Comment (jp9000) [reply libobs/graphics/effect-parser.c]: I think you still forgot this part
